### ISSUE TITLE
Race in ServiceLookup.CallSiteValidator.ValidateCallSite()

### ISSUE THREAD

**Issue by mikeharder** [aspnet/DependencyInjection#467]
@halter73, @pakrym 

I am getting the following exception when running the EF benchmarks (https://github.com/aspnet/benchmarks/issues/162):

```
System.ArgumentException: An item with the same key has already been added. Key: Benchmarks.Data.EfDb
   at System.ThrowHelper.ThrowAddingDuplicateWithKeyArgumentException(Object key)
   at System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteValidator.ValidateCallSite(Type serviceType, IServiceCallSite callSite) in C:\Git\DependencyInjection\src\Microsoft.Extensions.DependencyInjection\ServiceLookup\CallSiteValidator.cs:line 19
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.CreateServiceAccessor(Type serviceType, ServiceProvider serviceProvider) in C:\Git\DependencyInjection\src\Microsoft.Extensions.DependencyInjection\ServiceProvider.cs:line 76
   at System.Collections.Concurrent.ConcurrentDictionaryExtensions.GetOrAdd[TKey,TValue,TArg](ConcurrentDictionary`2 dictionary, TKey key, Func`3 valueFactory, TArg arg) in C:\Git\DependencyInjection\src\Microsoft.Extensions.DependencyInjection\Internal\ConcurrentDictionaryExtensions.cs:line 11
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.GetService(Type serviceType) in C:\Git\DependencyInjection\src\Microsoft.Extensions.DependencyInjection\ServiceProvider.cs:line 63
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetService[T](IServiceProvider provider) in C:\Git\DependencyInjection\src\Microsoft.Extensions.DependencyInjection.Abstractions\ServiceProviderServiceExtensions.cs:line 28
   at Benchmarks.Middleware.MultipleUpdatesEfMiddleware.<Invoke>d__4.MoveNext() in C:\Git\Benchmarks\src\Benchmarks\Middleware\MultipleUpdatesEfMiddleware.cs:line 0
```

The issue appears to be a race in ServiceLookup.CallSiteValidator.ValidateCallSite():

```
public void ValidateCallSite(Type serviceType, IServiceCallSite callSite)
{
    var scoped = VisitCallSite(callSite, default(CallSiteValidatorState));
    if (scoped != null)
    {
        _scopedServices.Add(serviceType, scoped);
    }
}
```

https://github.com/aspnet/DependencyInjection/blob/dev/src/Microsoft.Extensions.DependencyInjection/ServiceLookup/CallSiteValidator.cs#L15-L21

**Comment by henkmollema** [comment]
@pakrym @davidfowl in which release will this be fixed?

**Comment by pakrym** [comment]
@henkmollema this was not back ported to patch releases so I think this would release with 2.0

**Comment by henkmollema** [comment]
Ah, bummer. Thanks anyways @pakrym. 😄

**Comment by jdom** [comment]
I'm encountering the same issue. Is there a workaround to avoid the race for now? should I just disable scope validation or is there something less risky?

**Comment by jdom** [comment]
ah, I see this will be included in a patch release, even though it said here it would only go to 2.0.
Any ETA of when 1.1.2 will be released?

**Comment by henkmollema** [comment]
@jdom the packages have been released to NuGet. It's available in the 1.1.1 version: https://www.nuget.org/packages/Microsoft.Extensions.DependencyInjection/1.1.1.

**Comment by jdom** [comment]
Ah, great, thank you!
